lib.makeOverridable: Propagate function arguments

infinisil opened the pull request:
###### Motivation for this change

- Propagate the arguments of `f` to `(makeOverridable f args).override`

This allows people to query what arguments they can override for
derivations without having to look at the nix expression with

```
nix-repl> lib.functionArgs pkgs.hello.override
{ fetchurl = false; stdenv = false; }
```

- If `f args` is a function, propagates its arguments to
`makeOverridable f args`

This allows querying function arguments even after it's been
makeOverridable'd. E.g.

```
nix-repl> lib.functionArgs pkgs.fetchgit
{ branchName = true; deepClone = true; fetchSubmodules = true; ... }
```

This is a continuation of https://github.com/NixOS/nixpkgs/pull/47535, this time with correct behavior even when `.overrideDerivation` or `.overrideAttrs` is used.

I took the liberty to remove the `overrideDerivation = throw "overrideDerivation not yet supported for functors";` while I'm at it.

Ping @nbp @Profpatsch 

###### Things done

- [x] Double-checked to make sure the code doesn't change any behavior

infinisil committed:
gimp: Fix derivation detection

Infinisil commented on pkgs/applications/graphics/gimp/wrapper.nix:
```diff
@@ -1,7 +1,7 @@
 
 let
-allPlugins = lib.filter (pkg: builtins.isAttrs pkg && pkg.type == "derivation" && !pkg.meta.broken or false) (lib.attrValues gimpPlugins);
+allPlugins = lib.filter (pkg: lib.isDerivation pkg && !pkg.meta.broken or false) (lib.attrValues gimpPlugins);
```
This change is necessary because `gimpPlugins` was `makeOverride`'d, meaning it has a `.override`, which previously was a function, but now is a functor attrset (but without a `.type`), which this function wasn't able to handle before. The `lib.isDerivation` function properly checks for that.

FRidh commented:
Interesting change. Does this impact memory usage?

grahamc commented:
@GrahamcOfBorg eval

it should have posted the performance report ...

Infinisil commented:
Hm that doesn't look very good, why would eval time go up this much? I call bad luck, let's try again

@GrahamcOfBorg eval

| stat | before | after | Δ | Δ% |
| --- | --- | --- | --- | --- |
| cpuTime | 189.76 | 225.71 | ↗ 35.95 | 18.94% |
| envs-bytes | 3,013,206,344 | 3,022,604,696 | ↗ 9,398,352 | 0.31% |
| envs-elements | 162,003,077 | 162,976,653 | ↗ 973,576 | 0.60% |
| envs-number | 107,323,858 | 107,424,467 | ↗ 100,609 | 0.09% |
| gc-heapSize | 11,828,391,936 | 11,878,723,584 | ↗ 50,331,648 | 0.43% |
| gc-totalBytes | 23,316,522,048 | 23,352,482,976 | ↗ 35,960,928 | 0.15% |
| list-bytes | 1,591,287,664 | 1,591,287,664 | 0 |  |
| list-concats | 7,288,187 | 7,288,187 | 0 |  |
| list-elements | 198,910,958 | 198,910,958 | 0 |  |
| nrAvoided | 151,146,250 | 151,485,068 | ↗ 338,818 | 0.22% |
| nrFunctionCalls | 93,631,170 | 93,731,779 | ↗ 100,609 | 0.11% |
| nrLookups | 75,657,344 | 75,682,341 | ↗ 24,997 | 0.03% |
| nrOpUpdateValuesCopied | 230,487,746 | 230,496,380 | ↗ 8,634 | 0.00% |
| nrOpUpdates | 12,981,495 | 12,984,373 | ↗ 2,878 | 0.02% |
| nrPrimOpCalls | 66,132,513 | 66,132,513 | 0 |  |
| nrThunks | 155,344,587 | 155,998,851 | ↗ 654,264 | 0.42% |
| sets-bytes | 7,771,229,080 | 7,772,720,896 | ↗ 1,491,816 | 0.02% |
| sets-elements | 314,516,021 | 314,568,889 | ↗ 52,868 | 0.02% |
| sets-number | 27,855,572 | 27,883,445 | ↗ 27,873 | 0.10% |
| sizes-Attr | 24 | 24 | 0 |  |
| sizes-Bindings | 8 | 8 | 0 |  |
| sizes-Env | 16 | 16 | 0 |  |
| sizes-Value | 24 | 24 | 0 |  |
| symbols-bytes | 15,190,949 | 15,190,922 | ↘ 27 | -0.00% |
| symbols-number | 334,133 | 334,134 | ↗ 1 | 0.00% |
| values-bytes | 5,461,096,464 | 5,477,392,272 | ↗ 16,295,808 | 0.30% |
| values-number | 227,545,686 | 228,224,678 | ↗ 678,992 | 0.30% |

Infinisil commented:
Yeah looking much better now (only 3.5% slower), not sure how much I can trust the eval time now.

Profpatsch commented:
Can you split this PR into two/three/n commits, the implementation changes first and then any renaming/refactoring afterwards?

This is just a block of stuff and it’s very hard to see what is renaming, what is actual changes to the logic and what is reformatting/reflowing of the code.

infinisil committed:
lib/makeOverridable: Deduplicate override definition

And call it overrideArgs in the let binding because that's what it does

infinisil committed:
lib/makeOverridable: Abstract result overriding

infinisil committed:
lib/makeOverridable: Propagate function arguments to override functions

This allows querying the arguments you can .override:

  nix-repl> lib.functionArgs pkgs.hello.override
  { fetchurl = false; stdenv = false; }

infinisil committed:
lib/makeOverridable: Propagate function args of the callPackage'd function

This allows querying function arguments of things like fetchFromGitHub:

  nix-repl> lib.functionArgs pkgs.fetchFromGitHub
  { fetchSubmodules = true; githubBase = true; ... }

infinisil committed:
lib/makeOverridable: Remove unimplemented overrideDerivation for functions

- Apparently nobody ever needed this
- We already have enough ways to override things
- Using overrideDerivation is discouraged

infinisil committed:
lib/makeOverridable: Refactor

- Rename ff to result because that's what it is
- Better indentation
- Less parens
- Comment what overrideWith does

Infinisil commented:
@Profpatsch Yeah good point, did that, now best reviewable by commit

Infinisil commented:
@jonringer Got any concerns about this change?

jonringer commented:
No concerns code-wise. The inner-workings of nix looks like voodoo to me :)

For commit hygience, maybe squash a few of the minor ones together. Otherwise LGTM

Infinisil commented:
@jonringer (asking because you put a :confused: on the PR)

jonringer commented:
oh, I think i meant to comment about how I couldnt make heads or tails of what magic you were doing. Sorry for the confusion :)

Infinisil commented:
Hehe, this time in the performance report:


stat | before | after | Δ | Δ%
-- | -- | -- | -- | --
cpuTime | 192.73 | 192.73 | ↗ 0.00 | 0.00%

Profpatsch commented on lib/customisation.nix:
```diff
@@ -66,28 +66,31 @@ rec {
+        }
+      else if lib.isFunction result then
+        # Transform the result into a functor while propagating its arguments
+        lib.setFunctionArgs result (lib.functionArgs result) // {
```
You defined `copyArgs` in an earlier commit, so why not use it here as well?

Infinisil replied:
It's not the same, `copyArgs` makes a function have the args of `f`, while this is `result`

Profpatsch commented on lib/customisation.nix:
```diff
@@ -66,28 +66,31 @@ rec {
+          override = overrideArgs;
+          overrideDerivation = fdrv: overrideResult (x: overrideDerivation x fdrv);
+          ${if result ? overrideAttrs then "overrideAttrs" else null} = fdrv:
+            overrideResult (x: x.overrideAttrs fdrv);
```
I’d like to see a deprecation of `overrideDerivation`, by throwing a warning and removing it next release (or the one after).

Infinisil replied:
The attribute has been throwing an error already forever, so removing it is already the next step

Profpatsch commented on lib/customisation.nix:
```diff
@@ -66,28 +66,31 @@ rec {
+      result = f origArgs;
+
       # Creates a functor with the same arguments as f
       copyArgs = g: lib.setFunctionArgs g (lib.functionArgs f);
```
If `f` was already a functor, we now have multiple levels of nested attrsets with essentially the same information, right? Shouldn’t we check for that?

Infinisil replied:
Yeah, I just added a new commit that makes `setFunctionArgs` not do unnecessary nesting

Profpatsch reviewed (changes requested):
I found a few things.

Infinisil commented:
@Profpatsch Can you take another look?

Profpatsch commented on lib/trivial.nix:
```diff
@@ -277,10 +277,17 @@ rec {
+    else if builtins.isAttrs f && f ? __functor then
+      f // {
+        __functionArgs = args;
+      }
```
I’m not sure about the overwriting of existing `__functionArgs` here. What’s the intended semantics?

Infinisil replied:
Well the function is called `setFunctionArgs`, so it makes sense that any previous value is dicarded, like `set a . set b == set a` in Haskell land

Profpatsch replied:
Ah, let’s add it to the documentation then.

But then the question is actually why we might run into a situation where we call set multiple times in the first place (in the calling code).

Infinisil replied:
Hehe yes, that's why this function was added in the first place, see your own comment in https://github.com/NixOS/nixpkgs/pull/67809#discussion_r326875446

Profpatsch replied:
Yeah, but I’m pretty sure the interface is broken if calling `makeOverridable` multiple times overwrites the previous result.

It’s essentially the focus of this PR, so that’s a pretty important nitpick I’d say. :)

Infinisil replied:
Ah yeah you actually got something there. The case of a doubly-wrapped functor can never occur in `makeOverridable`!

The only potential candidates are the `setFunctionArgs` in `copyArgs`, which is only ever called with a newly constructed function (no functor possible), or in the `lib.isFunction result` case, but because `isAttrs` is true for functors too, it would have been caught in the previous condition already, therefore `result` can't be a functor either.

Infinisil replied:
So it really doesn't matter for this PR, I'll get rid of the `setFunctionArgs` double-wrap-prevention commit. It's just added complexity without any use

Profpatsch reviewed (changes requested):
Sorry for the wait, was on vacation for a week.

Infinisil commented:
@GrahamcOfBorg eval

Not sure why it failed evaluation, seems unrelated
